rust/agama-files/src/runner.rs

```diff
@@ -0,0 +1,365 @@
+
+    /// Runs the script.
+    ///
+    /// If the script fails, it asks the user whether it should try again.
+    async fn run_script(&self, script: &Script) {
+        loop {
+            let path = self
+                .workdir
+                .join(script.group().to_string())
+                .join(script.name());
+
+            let Err(error) = self.run_command(&path, script.chroot()).await else {
+                return;
+            };
+
+            if !self.should_retry(&script, error).await {
+                return;
+            }
+        }
+    }
+
+    ///  Asks the user whether it should try to run the script again.
+    async fn should_retry(&self, script: &Script, error: Error) -> bool {
+        let text = format!(
+            "Running the script '{}' failed. Do you want to try again?",
+            script.name()
+        );
+        let mut question = QuestionSpec::new(&text, "scripts.retry").with_yes_no_actions();
+
+        if let Error::Script(output) = error {
+            let stdout = String::from_utf8_lossy(&output.stdout);
+            let stderr = String::from_utf8_lossy(&output.stderr);
+            let exit_status = output
+                .status
+                .code()
+                .map(|c| c.to_string())
+                .unwrap_or("unknown".to_string());
+            question = question.with_data(&[
+                ("name", script.name()),
+                ("stdout", &stdout),
+                ("stderr", &stderr),
+                ("exit_status", &exit_status),
+            ]);
+        }
+
+        let answer = ask_question(&self.questions, question).await.unwrap();
+
+        return answer.action == "Yes";
+    }
+
+    /// Runs the script at the given path.
+    ///
+    /// * `path`: script's path.
+    /// * `chroot`: whether to run the script in a chroot.
+    async fn run_command<P: AsRef<Path>>(&self, path: P, chroot: bool) -> Result<(), Error> {
+        let path = path.as_ref();
+        let command = if chroot {
+            let mut command = process::Command::new("chroot");
+            command.args([&self.install_dir, path]);
+            command
+        } else {
+            process::Command::new(path)
+        };
+
+        let output = run_with_retry(command)
+            .await
+            .inspect_err(|e| println!("Error executing the script: {e}"))?;
+
+        fs::write(path.with_extension("log"), output.stdout.clone())?;
+        fs::write(path.with_extension("err"), output.stderr.clone())?;
+        if let Some(code) = output.status.code() {
+            fs::write(path.with_extension("out"), code.to_string())?;
+        }
+
+        if !output.status.success() {
+            return Err(Error::Script(output));
+        }
+
+        Ok(())
+    }
+
+    /// Ancillary function to start the progress.
+    fn start_progress(&self, scripts: &[&Script]) {
+        let messages: Vec<_> = scripts
+            .iter()
+            .map(|s| format!("Running user script '{}'", s.name()))
```

Comment:
small note about localization. If you change localization during progress..it won't switch localization when we have it working of course :)

Reply:
Yes, we need to go through the whole application when we have l10n/i18n working :-)